[Utility] trustless relays servicer token validation

adshmh opened the pull request:
## Description

Add validation of application's session tokens to the servicer.

## Issue

Part of work on #754 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

- Servicer now validates the availability of tokens in the current session before executing relays.

## Testing

- [x] `make develop_test`; if any code changes were made
- [ ] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [ ] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [ ] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [ ] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

adshmh committed:
merge main

adshmh committed:
Validate trustless relay: available service tokens for the application

adshmh committed:
merge main

adshmh committed:
Fix double error wrapping until we move to go 1.20

adshmh commented:
Will update this PR shortly to add/use Persistence module placeholders for maintaining state.

adshmh committed:
Feature: use persistence module for token usage maintenance

adshmh committed:
Add relay execution logic to servicer

adshmh commented:
Also pushed a commit to implement the relay execution.

Olshansk commented on utility/service/service.go:
```diff
@@ -70,17 +96,102 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+		return nil, fmt.Errorf("Error executing relay: %w", err)
+	}
+
+	// DISCUSS: should we validate the response from the node?
```
This is a big topic (data integrity, etc...)

I would add something like: `// TODO(M6): Look into data integrity checks and response validation.`

adshmh replied:
Thanks, done.

Olshansk commented on utility/service/service_test.go:
```diff
@@ -11,25 +15,53 @@ import (
-		relay    *coreTypes.Relay
-		expected error
+		name       string
+		usedTokens int64
```
s/usedTokens/usedSessionTokens

Is this `total` or `perRelay`?

adshmh replied:
Thank you for the review. Fixed.

Olshansk commented on utility/service/service_test.go:
```diff
@@ -43,63 +75,148 @@ func TestAdmitRelay(t *testing.T) {
 			expected: errValidateBlockHeight,
 		},
+		{
+			name:     "Relay not matching the servicer is rejected",
```
```suggestion
			name:     "Relay not matching the servicer in this session is rejected",
```

adshmh replied:
Thanks, fixed.

Olshansk commented on utility/service/service_test.go:
```diff
@@ -43,63 +75,148 @@ func TestAdmitRelay(t *testing.T) {
+			expected: errValidateServicer,
+		},
+		{
+			name:       "Relay for maxed-out app is rejected",
```
```suggestion
			name:       "Relay for app out of quota is rejected",
```

adshmh replied:
Thanks, fixed.

Olshansk commented on utility/service/service_test.go:
```diff
@@ -43,63 +75,148 @@ func TestAdmitRelay(t *testing.T) {
 			require.NoError(t, err)
 
-			err = servicerMod.(*servicer).admitRelay(testCase.relay)
+			servicer := servicerMod.(*servicer)
```
```suggestion
			servicer, ok := servicerMod.(*servicer)
			require.True(t, ok)
```

adshmh replied:
Thank you. Fixed.

Olshansk commented on utility/service/service.go:
```diff
@@ -70,17 +96,102 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+// hasCollision returns:
+//  1. The signed digest of a relay/response pair,
+//  2. Whether there was a collision for the specific chain (i.e. should the service proof be stored for claiming later)
+func (s *servicer) hasCollision(relay *coreTypes.Relay, response *coreTypes.RelayResponse) (digest []byte, collides bool, err error) {
```
s/collides/isReplay

adshmh replied:
Thank you for the review. I updated to `isRelayVolumeApplicable` as a proposed change since it seems to be the intention of the computation here, but please let me know if `isReplay` is a better fit.

Olshansk commented on utility/service/service.go:
```diff
@@ -70,17 +96,102 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+	return false, nil
+}
+
+func marshal(request *coreTypes.Relay, response *coreTypes.RelayResponse) ([]byte, error) {
```
s/marshal/marshalWhat?

1. Rename
s/marshal/marshalWhat?

2. Comment

Even though it's an internal helper, add a comment because this functionality is VERY important for the core of the protocol.

3. Refactor

The integrity of the blockchain (i.e. the protobufs and their serialized values in the tree) are the most important piece of the state. Ideally, people should be able to use the protos and implement it on their own. 

Can you create a proto in `relay.proto` that accepts both of these so that the marshaling of the digest is done by `codec.GetCodec().Marshal(RelayReqRes)`. That way we are language and framework agnostic and others, for example, could reimplement v1 in rust using the same protos.

```proto
message RelayReqRes {
	RelayReq
	RelayRes
}
```

adshmh replied:
Thank you for the review. Fixed.

Olshansk commented on utility/service/service.go:
```diff
@@ -70,17 +96,102 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+}
+
+// executeRelay performs the passed relay using an HTTP request to the chain-specific target URL
+func (s *servicer) executeRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse, error) {
```
Should we:

1. Rename to `executeHttpRelay`
2. Add a TODO to add support for others?
3. Add a switch-case inside to add support for other requests in the future?

adshmh replied:
Thank you for the review. Fixed.

Olshansk commented on utility/service/service.go:
```diff
@@ -122,19 +232,84 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
+		return fmt.Errorf("Error getting servicer token usage: application %s session number %d: %w", session.Application.PublicKey, session.SessionNumber, err)
+	}
+
+	if usedAppSessionTokens == nil || usedAppSessionTokens.Cmp(servicerAppSessionTokens) < 0 {
```
I think we should log a warning for `usedAppSessionTokens.Cmp(servicerAppSessionTokens) < 0`

adshmh replied:
Could you elaborate on this please? This condition is expected to occur whenever the application has sufficient tokens remaining for the session. Would a log message be used to track the relays?

Olshansk commented on utility/service/service.go:
```diff
@@ -70,17 +96,102 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+	}
+	defer writeCtx.Release()
+
+	// DISCUSS: should we extend/use UnitOfWork for updating/retrieving token usage?
```
This is a good question. `UnitOfWork` was meant to be more related to a single, atomic, rollbackable state transition related to on-chain things. A session is more of an "off chain" behaviour (spanning multiple blocks) that eventually transfers into a state transition.

Does that make sense? If so, can you add some of this context to the `UnitOfWork` interface?

adshmh replied:
It makes perfect sense. Thank you for the detailed explanation. Updated the PR and added a comment to `UtilityUnitOfWork` as a first draft.

Olshansk commented on utility/service/service.go:
```diff
@@ -70,17 +96,102 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+	defer writeCtx.Release()
+
+	// DISCUSS: should we extend/use UnitOfWork for updating/retrieving token usage?
+	if err := writeCtx.RecordRelayService(relay.Meta.ApplicationAddress, relayDigest, relay, response); err != nil {
```
IMPORTANT: @adshmh I unfortunately led you down the wrong path after our call yesterday.

- `readCtx` is used to read `on-chain` data (at a specific height)
	- We read on-chain data to understand the session quota (num tokens)

 -`writeCtx` is used to write to `on-chain` data (at a specific height)
	- We write on-chain data when the session is complete (e.g. need to send claim & proof transactions)

While the session is ongoing, we need to persist things (locally on disk or in memory but not on chain). Therefore, I think we might need to do something similar to `NonceDeduper` by leveraging `NewGenericFIFOSet` with a very large capacity.

I glossed over this but happy to go into more detail if need be. Just let me know.

adshmh replied:
@Olshansk Thank you for the detailed review. Considering the requirements described in the relay mining paper, use of SMT backed by the `kvstore` package seems a better fit here. 
Please advise whether this seems correct.

Olshansk replied:
> use of SMT backed by the kvstore package seems a better fit here (while also steering clear of writeCtx

Correct.

>  In contrast, NewGenericFIFISet makes sense if the intention is to build/update the SMT only once,

Got it, makes sense.

Note that there will be a new SMT instance for every session. It's how v0 works and how v1 works. 

Unlike the trees for the state, where there are N trees always growing/changing, there is one new tree for every session. If a node is handling a couple of session at once, it will maintain a couple new SMTs until those session are claimed & proven.

Does that make sense?

adshmh replied:
> Does that make sense?

Yes, thank you for the clarification.

> Unlike the trees for the state, where there are N trees always growing/changing, there is one new tree for every session. If a node is handling a couple of session at once, it will maintain a couple new SMTs until those session are claimed & proven.

Got it,  an SMT will be maintained for every session that has received service but has not been claimed yet.

Olshansk replied:
💯

h5law replied:
To add some context here the SMT writes are very efficient as we dont write to the database until `Commit()` is called so the writes are not very intensive until completed

Olshansk commented on utility/service/service.go:
```diff
@@ -122,19 +232,84 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
+	// This multiplication is performed to minimize the chance of under-utilization of application's tokens,
+	//	while removing the overhead of communication between servicers which would be necessary otherwise.
+	// see https://arxiv.org/abs/2305.10672 for details on application and servicer distributed rate-limiting
+	adjustedTokens := servicerTokens.Mul(servicerTokens, big.NewFloat(1+RelayAccuracyParameter))
```
Amazing 💯

Olshansk commented on utility/service/service.go:
```diff
@@ -122,19 +232,84 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
+	return s.totalTokens[session.Application.PublicKey]
+}
+
+// calculateServicerAppSessionTokens return the number of tokens the servicer can use for the application in the current session
```
```suggestion
// calculateServicerAppSessionTokens return the number of tokens the servicer has remaining for the Application in the session provided. If nothing is cached, the maximum number of session tokens is computed.
```

adshmh replied:
Thank you for fixing the comment. Updated.

adshmh replied:
Thank you for the review. Updated the comment.

Olshansk commented on utility/service/service.go:
```diff
@@ -122,19 +232,84 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
+}
+
+// calculateServicerAppSessionTokens return the number of tokens the servicer can use for the application in the current session
+func (s *servicer) calculateServicerAppSessionTokens(session *coreTypes.Session, currentHeight int64) (*big.Int, error) {
```
Why not use the `currentHeight` from the session?

adshmh replied:
Thank you for the review. Fixed.

adshmh replied:
Thank you for the review, fixed.

Olshansk commented on utility/service/service.go:
```diff
@@ -177,20 +351,44 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
 
+// DISCUSS: do we need to export this functionality as part of the utility module?
+// calculateAppSessionTokens determines the number of "session tokens" an application gets at the beginning
+// of every session. Each servicer will serve a maximum of (Session Tokens / Number of Servicers in the Session) relays for the application
```
```suggestion
// of every session. Each servicer will serve a maximum of ~(Session Tokens / Number of Servicers in the Session) relays for the application
```

adshmh replied:
Thank you for catching this. Fixed.

adshmh replied:
Thank you for catching this. Fixed.

Olshansk commented on utility/service/service.go:
```diff
@@ -350,3 +548,44 @@ func (r *relay) GetRelayChain() RelayChain              { return nil }
 func (r *relay) GetSignature() string                   { return "" }
+
+// executeHTTPRequest performs the HTTP request that sends the relay to the chain's URL.
+func executeHTTPRequest(cfg *configs.ChainConfig, relay *coreTypes.RelayPayload) (*coreTypes.RelayResponse, error) {
```
Not much to add here. If we point to the Pocket V1 blochcain & localhost instead of ETH (i.e. 0021) with a fake URL, it might be an even easier way to test things E2E.

adshmh replied:
Thank you, fixed.

Olshansk reviewed (changes requested):
@adshmh 

1. This is amazing work! 🙌 
2. Thank you for taking the initiative and working on this so quickly. There is more to come ;)
3. This almost in the right direction with the expectation of me leading you down the wrong path. Alongside all the comments, see the comment I left in `utility/service/service.go` and let me know if you have any more questions!

adshmh commented on utility/service/service.go:
```diff
@@ -70,17 +96,102 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+	return response, nil
+}
+
+// hasCollision returns:
```
Thank you for the review. Yes, `hasCollision` doesn't really say much. I updated to `isRelayVolumeApplicableOnChain` for now, as this is what the computation seems to intend to do, but please let me know if `isRelayReplay` is a better fit here.

adshmh commented on shared/core/types/proto/relay.proto:
```diff
@@ -10,6 +10,7 @@ message Relay {
     RelayMeta meta = 2;
 }
 
+// INCOMPLETE: add support for different payloads, e.g. JSON, GRPC, etc.
```
> I'm thinking we could potentially switch `data` from being `string` to being `bytes`

Sounds good to me: updated the `data` type to `bytes`.

> Would json & grpc payloads not be part of `data`?

Not sure about this one, so I kept this for the moment, but please let me know if it needs to be changed.
Two ways to approach this:
       1) Every different chain (or even service) has its own custom payload (e.g. HTTP, JSON, GRPC, non-chain services)
       2) The payloads are in the `[]byte` type data field.
I recommend the first approach as it provides more visibility to different pieces of the code that interact with a payload, but either approach would work.

adshmh commented on shared/modules/persistence_module.go:
```diff
@@ -129,6 +131,8 @@ type PersistenceWriteContext interface {
 	InitFlags() error
 	SetFlag(paramName string, value any, enabled bool) error
+
+	RecordRelayService(applicationAddress string, key []byte, relay *coreTypes.Relay, response *coreTypes.RelayResponse) error
```
> 1. Wdyt of this instead: `StoreServiceRelay`

Thank you for the review. Fixed.

>     2. What is `key` - please use a more self descriptive name

Thanks, fixed.
 
>     3. Optional: I think `appAddr` is sufficient but it's a mater of style and preference and doesn't matter if you prefer `applicationAddress`

Thanks, fixed.

adshmh committed:
Address review comments

adshmh commented on persistence/servicer.go:
```diff
@@ -79,3 +80,15 @@ func (p *PostgresContext) SetServicerPauseHeight(address []byte, height int64) e
+}
+
+// INCOMPLETE: implement this
+// GetServicerTokenUsage returns the number of tokens used by the servicer in the current session, i.e. for the application associated with the session
```
Thank you for the review. Fixed.

adshmh commented on persistence/servicer.go:
```diff
@@ -79,3 +80,15 @@ func (p *PostgresContext) SetServicerPauseHeight(address []byte, height int64) e
 }
+
+// INCOMPLETE: implement this
+// DISCUSS: both the relay and the response can be large structures: we may need to truncate the stored values
```
Thanks, fixed.

adshmh commented on runtime/configs/proto/utility_config.proto:
```diff
@@ -14,5 +14,18 @@ message UtilityConfig {
+  map<string, ChainConfig> chains = 3;
+}
+
+message ChainConfig {
```
Thanks, fixed.

adshmh commented on runtime/configs/proto/utility_config.proto:
```diff
@@ -14,5 +14,18 @@ message UtilityConfig {
   string public_key = 1;
   string address = 2;
-  repeated string chains = 3;
+  map<string, ChainConfig> chains = 3;
```
Thanks, fixed.

adshmh commented on runtime/configs/proto/utility_config.proto:
```diff
@@ -14,5 +14,18 @@ message UtilityConfig {
+
+message ChainConfig {
+  string url = 1;
+  string user_agent = 2;
```
Thank you for the review. Fixed.

adshmh commented on runtime/configs/proto/utility_config.proto:
```diff
@@ -14,5 +14,18 @@ message UtilityConfig {
+
+message BasicAuth {
+  string user_name = 1;
+  // DISCUSS: does this need to be encrypted?
```
Thanks for the review. Done.

adshmh commented on runtime/configs/proto/utility_config.proto:
```diff
@@ -14,5 +14,18 @@ message UtilityConfig {
+  map<string, ChainConfig> chains = 3;
+}
+
+message ChainConfig {
```
Thank you for the review. Fixed.

adshmh commented on runtime/docs/CHANGELOG.md:
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+## [0.0.0.41] - 2023-06-06
+
+- Add a new ChainConfig configuration field to servicer config
```
Thank you for the review. Fixed.

Olshansk replied:
Please resolve conversations that have been addressed in the future

adshmh commented on shared/modules/persistence_module.go:
```diff
@@ -219,4 +223,7 @@ type PersistenceReadContext interface {
 	GetBytesFlag(paramName string, height int64) ([]byte, bool, error)
+
+	// Servicer application tokens
+	GetServicerTokenUsage(session *coreTypes.Session) (*big.Int, error)
```
Thank you for the review. Fixed.

adshmh commented on runtime/configs/config.go:
```diff
@@ -121,7 +121,17 @@ func NewDefaultConfig(options ...func(*Config)) *Config {
 			ServicerConfig: &ServicerConfig{
-				Chains: []string{"0001"},
+				Chains: map[string]*ChainConfig{
+					"0021": {
```
Thank you for the review. Fixed.

adshmh commented on utility/service/service.go:
```diff
@@ -12,21 +21,38 @@ import (
+	typesUtil "github.com/pokt-network/pocket/utility/types"
 )
 
+// DISCUSS: where should the RelayAccracyParameter be defined?
```
Thank you, fixed.

adshmh commented on utility/service/service.go:
```diff
@@ -12,21 +21,38 @@ import (
 var (
 	errValidateBlockHeight = errors.New("relay failed block height validation")
 	errValidateRelayMeta   = errors.New("relay failed metadata validation")
+	errValidateServicer    = errors.New("relay does not match the servicer")
```
Thank you, fixed.

adshmh commented on utility/service/service.go:
```diff
@@ -12,21 +21,38 @@ import (
 )
 
+type sessionTokens struct {
+	SessionNumber int64
```
Thank you, fixed.

adshmh commented on utility/service/service.go:
```diff
@@ -12,21 +21,38 @@ import (
 )
 
+type sessionTokens struct {
+	SessionNumber int64
+	Count         *big.Int
```
Thanks. Fixed.

adshmh commented on utility/service/service.go:
```diff
@@ -12,21 +21,38 @@ import (
 	config *configs.ServicerConfig
+
+	rwlock sync.RWMutex
+	// totalTokens holds the total number of tokens assigned to this servicer for the app in the current session
```
Thanks, updated.

adshmh commented on utility/service/service.go:
```diff
@@ -12,21 +21,38 @@ import (
+
+	rwlock sync.RWMutex
+	// totalTokens holds the total number of tokens assigned to this servicer for the app in the current session
+	// DISCUSS: considering the computational complexity, should we skip caching this value?
```
Thanks, updated.

adshmh committed:
Merge remote-tracking branch 'origin/main' into feat-utility-trustless-relays-servicer-token-validation

adshmh committed:
Fix proto files

adshmh committed:
Remove gomock temp program

adshmh committed:
Fix linter errors

adshmh committed:
Fix failing unit test in runtime

Olshansk commented on utility/servicer/module_test.go:
```diff
@@ -11,25 +15,53 @@ import (
-	PublicKey: "a6cd0a304c38d76271f74dd3c90325144425d904ef1b9a6fbab9b201d75a998b",
-	Chains:    []string{"0021"},
-}
+const (
```
Can you create a ticket to add extensive E2E tests for this. See e2e/tests/root.feature

adshmh replied:
Done. #833

Olshansk commented on utility/servicer/module.go:
```diff
@@ -138,19 +235,86 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
-	*/
-	return nil
+func (s *servicer) validateApplication(session *coreTypes.Session) error {
+	// IMPROVE: use a function to get current height from the current session
```
1. Can't we just get it directly from `Session`?
2. If not, we should be able to do `s.GetBus().GetConsensusModule().CurrentHeight()`
3. If neither 1 or 2 work, this is `TECHDEBT`, not `IMPROVE` IMO

adshmh replied:
Thank you, fixed: the called function now gets the height from the session.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -138,19 +235,86 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
 }
 
-// TODO: implement
 // validateApplication makes sure the application has not received more relays than allocated in the current session.
```
```suggestion
// validateApplication makes sure the application has not received more relays than allocated in the current session.
// returns nil if the servicer should attempt to mine another relay for the session provided
```

adshmh replied:
Thank you, fixed.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -138,19 +235,86 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
-		}
-	*/
-	return nil
+func (s *servicer) validateApplication(session *coreTypes.Session) error {
```
I feel like we shold rename `validateApplication` to `sholdMineRelay`. Optionally returning a boolean or just using a nil error to signify `yes`

adshmh replied:
Thank you, fixed the name. Had to keep the return type as error: even though a boolean is preferable here, the `shouldMineRelay` function could run into errors: so just returning nil would indicate it is OK to mine.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -138,19 +235,86 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
+	}
+
+	if usedAppSessionTokens == nil || usedAppSessionTokens.Cmp(servicerAppSessionTokens) < 0 {
+		return nil
```
```suggestion
		return nil // should attempt to mine a relay
```

adshmh replied:
Thank you, fixed.

Olshansk commented on utility/servicer/module_test.go:
```diff
@@ -11,25 +15,53 @@ import (
+	}
+
+	testApp1 = &coreTypes.Actor{
+		Address:      "98a792b7aca673620132ef01f50e62caa58eca83",
```
Why not specify the actor type here?

adshmh replied:
Thanks, fixed.

Olshansk commented on utility/servicer/module_test.go:
```diff
@@ -11,25 +15,53 @@ import (
+		ActorType: coreTypes.ActorType_ACTOR_TYPE_SERVICER,
+		Address:   "a3d9ea9d9ad9c58bb96ec41340f83cb2cabb6496",
+		PublicKey: "a6cd0a304c38d76271f74dd3c90325144425d904ef1b9a6fbab9b201d75a998b",
+		Chains:    []string{"0021"},
```
```suggestion
		Chains:    []string{"POKT-UnitTestNet"},
```

adshmh replied:
Thank you, fixed.

Olshansk commented on utility/servicer/module_test.go:
```diff
@@ -11,25 +15,53 @@ import (
+	testServicer1 = &coreTypes.Actor{
+		ActorType: coreTypes.ActorType_ACTOR_TYPE_SERVICER,
+		Address:   "a3d9ea9d9ad9c58bb96ec41340f83cb2cabb6496",
+		PublicKey: "a6cd0a304c38d76271f74dd3c90325144425d904ef1b9a6fbab9b201d75a998b",
```
How did you generate these?

We have utilities to generate keys that are guaranteed to be compatible.

```go
	operatorKey, err := crypto.GeneratePublicKey()
	if err != nil {
		return nil, err
	}

	outputAddr, err := crypto.GenerateAddress()
	if err != nil {
		return nil, err
	}
```

adshmh replied:
Thank you for pointing this out. Fixed.

Olshansk commented on utility/servicer/module_test.go:
```diff
@@ -11,25 +15,53 @@ import (
+	}
+)
 
 func TestAdmitRelay(t *testing.T) {
```
```suggestion
func TestRelay_Admit(t *testing.T) {
```

adshmh replied:
Thanks, fixed.

Olshansk commented on utility/servicer/module_test.go:
```diff
@@ -43,63 +75,151 @@ func TestAdmitRelay(t *testing.T) {
 
-func testRelay(chain string, height int64) *coreTypes.Relay {
-	return &coreTypes.Relay{
+func TestExecuteRelay(t *testing.T) {
```
```suggestion
func TestRelay_Execute(t *testing.T) {
```

adshmh replied:
Thanks, fixed.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -138,19 +235,86 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
+func (s *servicer) calculateServicerAppSessionTokens(session *coreTypes.Session) (*big.Int, error) {
+	tokens := s.cachedAppTokens(session)
+	if tokens != nil && tokens.originalCountAvailable != nil && tokens.sessionNumber == session.SessionNumber {
+		return big.NewInt(1).Set(tokens.originalCountAvailable), nil
```
This returns the original count available but the comment says "number of tokens remaining".

Either the comment or business logic is off. Might also be worth updating the function name to what this is intended to do.

adshmh replied:
Thank you, fixed both the comment and the function name.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
+		return fmt.Errorf("%s: %w", err.Error(), errValidateRelayMeta)
 	}
 
 	// TODO: update the CLI to include ApplicationAddress(or Application Public Key) in the RelayMeta
```
Seems like this TODO is no longer relevant? If not, remove.

adshmh replied:
Thanks, fixed.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
+		return err
 	}
 
 	// TODO: (REFACTOR) use a loop to run all validators: would also remove the need for passing the session around
```
Seems like this TODO is no longer relevant? If not, remove.

adshmh replied:
Thanks, fixed.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
 	return nil
 }
 
+// DISCUSS: do we need to export this functionality as part of the utility module?
```
I think we can remove this `DISCUSS` item unless you see a reason for it. 

My answer is no.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
+		return nil, fmt.Errorf("Error processing application's staked amount %s: %w", session.Application.StakedAmount, coreTypes.ErrStringToBigInt(err))
+	}
+
+	// DOCUMENT: find the right document to explain the following:
```
```suggestion
	// TODO(M5): find the right document to explain the following:
```

adshmh replied:
Thanks, fixed.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
+	}
+	defer readCtx.Release() //nolint:errcheck // We only need to make sure the readCtx is released
+
+	// DISCUSS: using an interface for returning each defined parameter seems less error-prone: also could return e.g. int64 in this case to remove the type cast
```
Have you looked at the `GetParameter` function?

adshmh replied:
Thanks, that solved it. Removed the `DISCUSS` item.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
+	return appStake.Mul(appStake, big.NewInt(int64(appStakeTokensMultiplier))), nil
+}
+
+// executeHTTPRequest performs the HTTP request that sends the relay to the chain's URL.
```
```suggestion
// executeHTTPRequest performs the HTTP request that sends the relay to the chain's/service's URL.
```

adshmh replied:
Thanks, fixed.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
+// DISCUSS: do we need to export this functionality as part of the utility module?
+// calculateAppSessionTokens determines the number of "session tokens" an application gets at the beginning
+// of every session. Each servicer will serve a maximum of ~(Session Tokens / Number of Servicers in the Session) relays for the application
+func (s *servicer) calculateAppSessionTokens(session *coreTypes.Session) (*big.Int, error) {
```
```suggestion
// ADDTEST: Need to add more unit tests for the numerical portion of this functionality
func (s *servicer) calculateAppSessionTokens(session *coreTypes.Session) (*big.Int, error) {
```

adshmh replied:
Thanks, added.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -138,19 +235,86 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
+// calculateServicerAppSessionTokens return the number of tokens the servicer has remaining for the Application in the session provided.
+//
+//	If nothing is cached, the maximum number of session tokens is computed.
+func (s *servicer) calculateServicerAppSessionTokens(session *coreTypes.Session) (*big.Int, error) {
```
```suggestion
// ADDTEST: Need to add more unit tests for the numerical portion of this functionality
func (s *servicer) calculateServicerAppSessionTokens(session *coreTypes.Session) (*big.Int, error) {
```

adshmh replied:
Thanks, added.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -138,19 +235,86 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
-		}
-	*/
-	return nil
+func (s *servicer) validateApplication(session *coreTypes.Session) error {
```
```suggestion
// ADDTEST: Need to add more unit tests to account for potential edge cases
func (s *servicer) validateApplication(session *coreTypes.Session) error {
```

Olshansk commented on utility/servicer/module.go:
```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
+		return nil, fmt.Errorf("Chain %s not found in servicer configuration: %w", meta.RelayChain.Id, errValidateRelayMeta)
+	}
+
+	chainUrl, err := url.Parse(serviceConfig.Url)
```
```suggestion
	serviceUrl, err := url.Parse(serviceConfig.Url)
```

adshmh replied:
Thank you, fixed.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing chain URL %s: %w", serviceConfig.Url, err)
+	}
+	targetUrl := chainUrl.JoinPath(payload.HttpPath)
```
Can you add a comment on what this is doing? A small example in the comment above would help

adshmh replied:
Thanks for the review. This is not needed anymore: we have removed `HttpPath` from `JSONRPC` relays' payloads. Also fixed an issue with the HTTP call below: it should use POST: the `Method` on the JSONRPC is not an HTTP method.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
+	}
+	targetUrl := chainUrl.JoinPath(payload.HttpPath)
+
+	req, err := http.NewRequest(payload.Method, targetUrl.String(), bytes.NewBuffer([]byte(payload.Data)))
```
Should we convert `payload.Data` to `[]byte` from `string`? Doesn't have to be in this PR but maybe a TODO

adshmh replied:
Thank you for the review: This has been updated as per my comment above: the serialized form of `JSONRPC` payload is now being used.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
+		req.SetBasicAuth(serviceConfig.BasicAuth.UserName, serviceConfig.BasicAuth.Password)
+	}
+
+	// DISCUSS: do we need a default user-agent for HTTP requests?
```
```suggestion
	// INVESTIGATE: do we need a default user-agent for HTTP requests?
```

I don't know yet. If you have any insight/experience, can you add it in the TODO comment?

adshmh replied:
Thanks, fixed. Nothing to add yet.

Olshansk commented on utility/servicer/module.go:
```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
+	for k, v := range payload.Headers {
+		req.Header.Set(k, v)
+	}
+	if len(payload.Headers) == 0 {
```
Should this be `if "Content-Type" is not a key in `payload.Headers?`

adshmh replied:
Thank you for the review. Fixed.

Olshansk commented on utility/servicer/module_test.go:
```diff
@@ -43,63 +75,151 @@ func TestAdmitRelay(t *testing.T) {
+
+			servicer := &servicer{config: &config}
+			_, err := servicer.executeRelay(testCase.relay)
+			if !errors.Is(err, testCase.expectedErr) {
+				t.Fatalf("Expected error %v got: %v", testCase.expectedErr, err)
+			}
```
Why not use: https://pkg.go.dev/github.com/stretchr/testify/assert#ErrorIs

adshmh replied:
Thank you for catching this. Fixed.

Olshansk reviewed (changes requested):
@adshmh Great work!!

I did my best to go over everything but glossed over some parts of the PR.

adshmh commented on persistence/local.go:
```diff
@@ -0,0 +1,29 @@
+	databasePath string
+}
+
+// INCOMPLETE: implement this
```
done: #826 created and added to the INCOMPLETE item.

adshmh commented on shared/core/types/proto/relay.proto:
```diff
@@ -6,17 +6,26 @@ option go_package = "github.com/pokt-network/pocket/shared/core/types";
 
 // TECHDEBT: These structures were copied as placeholders from v0 and need to be updated to reflect changes in v1
```
Thank you, fixed.

adshmh commented on shared/core/types/proto/relay.proto:
```diff
@@ -6,17 +6,26 @@ option go_package = "github.com/pokt-network/pocket/shared/core/types";
 
-message RelayPayload {
-    string data = 1;
+// INCOMPLETE: add support for different payloads, e.g. JSON, GRPC, etc.
```
Thank you, updated and moved the comment.

adshmh commented on shared/core/types/proto/relay.proto:
```diff
@@ -6,17 +6,26 @@ option go_package = "github.com/pokt-network/pocket/shared/core/types";
-    string data = 1;
+// INCOMPLETE: add support for different payloads, e.g. JSON, GRPC, etc.
+message JsonRpcPayload {
+    bytes data = 1;
```
Thanks, added a commentn to `relay_payload` field.

adshmh commented on app/docs/CHANGELOG.md:
```diff
@@ -7,7 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
-## [0.0.0.8] - 2023-06-06
+## [0.0.0.8] - 2023-06-12
```
Thanks, fixed.

adshmh commented on build/docs/CHANGELOG.md:
```diff
@@ -7,7 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
-## [0.0.0.46] - 2023-06-06
+## [0.0.0.46] - 2023-06-12
```
Thanks, fixed.

adshmh commented on charts/CHANGELOG.md:
```diff
@@ -7,7 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
-## [0.0.0.6] - 2023-06-06
+## [0.0.0.6] - 2023-06-12
```
Thanks, fixed.

adshmh commented on shared/core/types/proto/relay.proto:
```diff
@@ -42,3 +51,8 @@ message Identifiable {
     string name = 2;
 }
+
+message RelayReqRes {
```
Thank you, fixed.

adshmh commented on consensus/doc/CHANGELOG.md:
```diff
@@ -7,7 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
-## [0.0.0.53] - 2023-06-08
+## [0.0.0.53] - 2023-06-12
```
Thanks, fixed.

adshmh commented on shared/modules/doc/CHANGELOG.md:
```diff
@@ -7,7 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
-## [0.0.0.14] - 2023-06-06
+## [0.0.0.14] - 2023-06-12
```
Thanks, fixed.

adshmh commented on shared/modules/persistence_module.go:
```diff
@@ -18,6 +20,9 @@ type PersistenceModule interface {
 
 	// Context operations
 	NewRWContext(height int64) (PersistenceRWContext, error)
+	// DISCUSS: removing height from "NewReadContext" input and passing it to specific methods seems a better choice.
```
Thank you. Done and done.

adshmh commented on shared/modules/persistence_module.go:
```diff
@@ -220,3 +232,23 @@ type PersistenceReadContext interface {
+	//	b) prepare claim/proof messages once the session is over
+	// The "relayDigest" and "relayReqResBytes" parameters will be used as key and leaf contents in the constructed SMT, respectively.
+	// OPTIMIZE: both the relay and the response can be large structures: we may need to truncate the stored values
+	StoreServiceRelay(session *coreTypes.Session, appAddr string, relayDigest, relayReqResBytes []byte) error
```
Thank you for catching this. Fixed.

adshmh commented on shared/modules/persistence_module.go:
```diff
@@ -220,3 +232,23 @@ type PersistenceReadContext interface {
+	//	b) prepare claim/proof messages once the session is over
+	// The "relayDigest" and "relayReqResBytes" parameters will be used as key and leaf contents in the constructed SMT, respectively.
+	// OPTIMIZE: both the relay and the response can be large structures: we may need to truncate the stored values
+	StoreServiceRelay(session *coreTypes.Session, appAddr string, relayDigest, relayReqResBytes []byte) error
```
Thanks, reads better with the update. Fixed.

adshmh commented on shared/modules/persistence_module.go:
```diff
@@ -220,3 +232,23 @@ type PersistenceReadContext interface {
+	// The "relayDigest" and "relayReqResBytes" parameters will be used as key and leaf contents in the constructed SMT, respectively.
+	// OPTIMIZE: both the relay and the response can be large structures: we may need to truncate the stored values
+	StoreServiceRelay(session *coreTypes.Session, appAddr string, relayDigest, relayReqResBytes []byte) error
+	// GetSessionTokensUsed returns the number of tokens that have been used for the passed session.
```
Thanks, fixed.

adshmh commented on shared/modules/persistence_module.go:
```diff
@@ -220,3 +232,23 @@ type PersistenceReadContext interface {
+	// OPTIMIZE: both the relay and the response can be large structures: we may need to truncate the stored values
+	StoreServiceRelay(session *coreTypes.Session, appAddr string, relayDigest, relayReqResBytes []byte) error
+	// GetSessionTokensUsed returns the number of tokens that have been used for the passed session.
+	//    This gets calculated from local storage, so it returns the count of tokens used by the servicer instance
```
Makes sense. Thanks. Fixed.

adshmh commented on shared/modules/persistence_module.go:
```diff
@@ -220,3 +232,23 @@ type PersistenceReadContext interface {
+	//	for the application associated with the session
+	GetSessionTokensUsed(*coreTypes.Session) (*big.Int, error)
+	// Release the allocated local context.
+	Release() error
```
Thanks, great point. Fixed. Place-holders are useful, but clearly easier to make mistakes on.

adshmh commented on shared/modules/persistence_module.go:
```diff
@@ -220,3 +232,23 @@ type PersistenceReadContext interface {
+	//	a) check the number of tokens used per session, and
+	//	b) prepare claim/proof messages once the session is over
+	// The "relayDigest" and "relayReqResBytes" parameters will be used as key and leaf contents in the constructed SMT, respectively.
+	// OPTIMIZE: both the relay and the response can be large structures: we may need to truncate the stored values
```
Thanks, fixed.

adshmh commented on shared/modules/persistence_module.go:
```diff
@@ -31,6 +36,11 @@ type PersistenceModule interface {
+	// NewLocalContext returns a local persistence context that can be used to store/retrieve node-specific, i.e. off-chain, data
+	// The module can maintain a single local context for both read and write operations: subsequent calls to NewLocalContext return
+	// the same local context.
+	NewLocalContext() (PersistenceLocalContext, error)
```
Thanks, fixed.

adshmh commented on runtime/configs/proto/persistence_config.proto:
```diff
@@ -16,4 +16,5 @@ message PersistenceConfig {
   string max_conn_lifetime = 8; // See pkg.go.dev/time#ParseDuration for reference
   string max_conn_idle_time = 9; // See pkg.go.dev/time#ParseDuration for reference
   string health_check_period = 10; // See pkg.go.dev/time#ParseDuration for reference
+  string local_database_path = 11; // The path used to store local, i.e. off-chain and node-specific, data.
```
Thanks, fixed.

adshmh commented on runtime/configs/proto/servicer_config.proto:
```diff
@@ -5,11 +5,33 @@ package configs;
+  double relay_mining_volume_accuracy = 5;
+}
+
+message ServiceConfig {
```
Thanks, the suggested reads very well, so I just copied it as a new comment.

adshmh commented on utility/servicer/module.go:
```diff
@@ -1,20 +1,34 @@
 	"golang.org/x/exp/slices"
 )
 
 var (
```
Thanks, updated.

adshmh commented on utility/servicer/module.go:
```diff
@@ -23,12 +37,25 @@ const (
 )
 
+// sessionTokens is used to cache the original number of tokens available
+// during a specific session: it is used as the value for a map with keys being applications' public keys
```
Thanks, added a TODO.

adshmh commented on utility/servicer/module.go:
```diff
@@ -23,12 +37,25 @@ const (
+// during a specific session: it is used as the value for a map with keys being applications' public keys
+type sessionTokens struct {
+	sessionNumber          int64
+	originalCountAvailable *big.Int
```
Reads better, thanks. Updated.

adshmh commented on utility/servicer/module.go:
```diff
@@ -23,12 +37,25 @@ const (
+	// This lock is needed to allow multiple GO routines update the totalTokens cache as part of serving relays
+	rwlock sync.RWMutex
+	// totalTokens is a mapping from application public keys to session metadata to keep track of session tokens
+	// INVESTIGATE: considering the computational complexity, should we skip caching this value?
```
Good point. Not sure how to phrase this. The idea is to avoid calling GetParam and performing `math.big` operations on every single relay.

Olshansk replied:
`OPTIMIZE: There is an opportunity to simplify the code through various means such as, but not limited to, avoiding extra math.big operations or excess GetParam calls`

How about this?

adshmh replied:
Looks great, updated. Thanks.

adshmh commented on utility/servicer/module.go:
```diff
@@ -23,12 +37,25 @@ const (
 	config *configs.ServicerConfig
+
+	// This lock is needed to allow multiple GO routines update the totalTokens cache as part of serving relays
+	rwlock sync.RWMutex
```
I suggest (but not strongly) to keep the map + mutex. We will very likely need to maintain more state in the servicer. In addition, `sync.Map` doc seems to say it is not optimized for our use-case:

>The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, >for better type safety and to make it easier to maintain other invariants along with the map content.

>The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once >but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite >entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention >compared to a Go map paired with a separate Mutex or RWMutex. `

Please let me know if using a `sync.Map` seems a better fit here.

Olshansk replied:
Add a comment: `NB: per the description in pkg.go.dev/sync#Map, we have chosen explicitly not to use sync.Map`

adshmh replied:
Thank you, updated.

adshmh commented on persistence/local.go:
```diff
@@ -0,0 +1,29 @@
+
+	coreTypes "github.com/pokt-network/pocket/shared/core/types"
+)
+
```
Thanks, fixed.

adshmh commented on runtime/configs/proto/utility_config.proto:
```diff
@@ -4,7 +4,10 @@ package configs;
 message UtilityConfig {
   uint64 max_mempool_transaction_bytes = 1;
   uint32 max_mempool_transactions = 2;
+  ServicerConfig servicer_config = 3;
```
Thanks, fixed.

adshmh commented on persistence/local.go:
```diff
@@ -0,0 +1,29 @@
+
+// INCOMPLETE: implement this
+// StoreServiceRelay implements the PersistenceLocalContext interface
+func (local *persistenceLocalContext) StoreServiceRelay(session *coreTypes.Session, appAddr string, relayDigest, relayReqResBytes []byte) error {
```
Thanks, fixed.

adshmh commented on persistence/local.go:
```diff
@@ -0,0 +1,29 @@
+	return nil
+}
+
+// INCOMPLETE: implement this
```
Thanks, fixed. #826

adshmh commented on utility/servicer/module.go:
```diff
@@ -86,17 +113,88 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+		return nil, fmt.Errorf("Error recording service proof for application %s: %w", relay.Meta.ApplicationAddress, err)
+	}
+
+	if err := localCtx.Release(); err != nil {
```
Thanks, fixed: `Release` has been removed as per a previous review comment.

adshmh commented on persistence/local.go:
```diff
@@ -0,0 +1,29 @@
+	coreTypes "github.com/pokt-network/pocket/shared/core/types"
+)
+
+type persistenceLocalContext struct {
```
Thank you for the review. Fixed.

adshmh commented on utility/servicer/module.go:
```diff
@@ -86,17 +113,88 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+}
+
+// isRelayVolumeApplicable returns:
+//  1. The signed digest of a relay/response pair,
```
Thanks, fixed.

adshmh commented on utility/servicer/module.go:
```diff
@@ -86,17 +113,88 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+
+// isRelayVolumeApplicable returns:
+//  1. The signed digest of a relay/response pair,
+//  2. Whether there was a collision for the specific chain (i.e. should the service proof be stored for claiming later)
```
Thank you for pointing this out. Looking at this again, I would appreciate some clarification:

There are 2 separate items to consider here:
1. Is the relay a replay? i.e. has the servicer serviced this exact same relay before?
2. Is a legit, i.e. not-replay, relay eligible for claiming rewards?

The current code doesn't try to address 1, but item 2, as per my understanding of the relay-mining paper:
Once a relay has been served, the servicer will check to see if the relay is eligible for claiming a reward based on the difficulty set for the chain/service: this can not be known in-advance, as it depends on the relay's response.

Olshansk replied:
> Is the relay a replay? i.e. has the servicer serviced this exact same relay before?

This is handled by default through the SMT algorithm. Won't go into the details here but lmk if we need to discuss it. It's explained in the Relay Mining paper

> Is a legit, i.e. not-replay, relay eligible for claiming rewards?

Correct. Including the proper signature.

> The current code doesn't try to address 1,

It's addressed by the algorithm. Out of scope of this PR but when we work on rate limiting and byzaneint cases, we can go into it.

> Whether there was a collision for the specific chain

It seems like your comment (1) and (2) are valid points but are not related to the comment in the code.

I still do not know what you mean by `Whether there was a collision for the specific chain`

adshmh commented on utility/servicer/module.go:
```diff
@@ -86,17 +113,88 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+// isRelayVolumeApplicable returns:
+//  1. The signed digest of a relay/response pair,
+//  2. Whether there was a collision for the specific chain (i.e. should the service proof be stored for claiming later)
+func (s *servicer) isRelayVolumeApplicable(relay *coreTypes.Relay, response *coreTypes.RelayResponse) (digest, serializedRelayRes []byte, collides bool, err error) {
```
Thanks, renamed to `doesCollide` to stay consistent with my comment above. But this may need to be renamed depending on your answer to the replay vs. legit-but-not reward-eligible relays.

adshmh commented on utility/servicer/module.go:
```diff
@@ -86,17 +113,88 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("Error marshalling relay and/or response: %w", err)
+	}
+
+	relayDigest := crypto.SHA3Hash(relayReqResBytes)
+
```
Thanks, fixed.

adshmh commented on utility/servicer/module.go:
```diff
@@ -86,17 +113,88 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+
+	signedDigest := s.sign(relayDigest)
+	response.ServicerSignature = hex.EncodeToString(signedDigest)
+	collision, err := s.isRelayVolumeApplicableOnChain(relayDigest, relay.Meta.RelayChain.Id)
```
Thank you, fixed.

adshmh commented on utility/servicer/module.go:
```diff
@@ -86,17 +113,88 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+	response.ServicerSignature = hex.EncodeToString(signedDigest)
+	collision, err := s.isRelayVolumeApplicableOnChain(relayDigest, relay.Meta.RelayChain.Id)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("Error checking collision for chain %s: %w", relay.Meta.RelayChain.Id, err)
```
Thank you, fixed.

adshmh commented on rpc/handlers.go:
```diff
@@ -114,8 +114,8 @@ func (s *rpcServer) PostV1ClientRelay(ctx echo.Context) error {
-	payload := &coreTypes.RelayPayload{
-		Data:     body.Payload.Data,
+	payload := &coreTypes.JsonRpcPayload{
+		Data:     []byte(body.Payload.Data),
```
This was from a missing `format` instruction on the openapi spec. Fixed.

adshmh commented on runtime/configs/config.go:
```diff
@@ -211,3 +213,21 @@ func CreateTempConfig(cfg *Config) (*Config, error) {
+		RelayMiningVolumeAccuracy: 0.2,
+		Services: map[string]*ServiceConfig{
+			// DISCUSS: what should the default config be? POKT/ETHM?
+			"POKT-V1": {
```
Thanks, fixed.

adshmh commented on runtime/configs/config.go:
```diff
@@ -211,3 +213,21 @@ func CreateTempConfig(cfg *Config) (*Config, error) {
+		Enabled:                   true,
+		RelayMiningVolumeAccuracy: 0.2,
+		Services: map[string]*ServiceConfig{
+			// DISCUSS: what should the default config be? POKT/ETHM?
```
Thank you, done: #831

adshmh commented on runtime/configs/config.go:
```diff
@@ -124,6 +125,7 @@ func NewDefaultConfig(options ...func(*Config)) *Config {
 			},
 		},
 		Utility: &UtilityConfig{
+			ServicerConfig:             defaultServicerConfig(),
```
Thank you for the review. Fixed.

adshmh commented on utility/servicer/module.go:
```diff
@@ -86,17 +113,88 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+	return signedDigest, relayReqResBytes, collision, nil
+}
+
+// INCOMPLETE: implement this
```
Thank you, fixed: #832

adshmh commented on utility/servicer/module.go:
```diff
@@ -86,17 +113,88 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+}
+
+// INCOMPLETE: implement this
+func (s *servicer) isRelayVolumeApplicableOnChain(digest []byte, relayChainId string) (bool, error) {
```
Thank you for catching this, fixed.

adshmh commented on utility/servicer/module_test.go:
```diff
@@ -43,63 +75,151 @@ func TestAdmitRelay(t *testing.T) {
 		})
 	}
 }
 
```
Thank you, fixed.

adshmh commented on utility/servicer/module.go:
```diff
@@ -138,19 +235,86 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
+		session.SessionNumber)
+}
+
+func (s *servicer) cachedAppTokens(session *coreTypes.Session) *sessionTokens {
```
Thanks, fixed.

adshmh commented on utility/servicer/module.go:
```diff
@@ -138,19 +235,86 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
+	defer s.rwlock.Unlock()
+
+	if len(s.totalTokens) == 0 {
+		s.totalTokens = make(map[string]*sessionTokens)
```
Thank you, fixed.

adshmh commented on utility/servicer/module_test.go:
```diff
@@ -43,63 +75,151 @@ func TestAdmitRelay(t *testing.T) {
+			if !errors.Is(err, testCase.expectedErr) {
+				t.Fatalf("Expected error %v got: %v", testCase.expectedErr, err)
+			}
+			// INCOMPLETE: verify HTTP request properties: payload/headers/user-agent/etc.
```
Thanks, fixed.

adshmh commented on utility/servicer/module_test.go:
```diff
@@ -43,63 +75,151 @@ func TestAdmitRelay(t *testing.T) {
+			name:              "Relay for app out of quota is rejected",
+			relay:             testRelay(),
+			usedSessionTokens: 999999,
+			expected:          errValidateApplication,
```
Thank you for the review. Fixed.

adshmh commented on utility/servicer/module.go:
```diff
@@ -182,31 +356,107 @@ func (s servicer) admitRelay(relay *coreTypes.Relay) error {
+		req.Header.Set("Content-Type", "application/json")
+	}
+
+	// DISCUSS: we need to optimize usage of HTTP client, e.g. for connection reuse, considering the expected volume of relays
```
Thank you for the review. Opened #837 and updated the comment.

adshmh committed:
Address review comments round 2

adshmh committed:
Merge remote-tracking branch 'origin/main' into feat-utility-trustless-relays-servicer-token-validation

adshmh committed:
Fix linter warnings

adshmh committed:
Fix default configuration override

adshmh commented:
@Olshansk the review comments are now addressed, except a few that may need further discussion. I also added a commit to improve the default config behavior, as the introduction of the servicer default config combined with incorrect default config behavior was breaking some of the unit tests.

Olshansk commented:
@adshmh regarding the three outstanding comments, please see my replies:
- https://github.com/pokt-network/pocket/pull/803#discussion_r1228679186
- https://github.com/pokt-network/pocket/pull/803#discussion_r1228679364
- https://github.com/pokt-network/pocket/pull/803#discussion_r1228705803

Will take a look at the changes shortly

adshmh committed:
Address review comments

adshmh commented:
> @adshmh regarding the three outstanding comments, please see my replies:
> 
>     * [[Utility] trustless relays servicer token validation #803 (comment)](https://github.com/pokt-network/pocket/pull/803#discussion_r1228679186)
> 
>     * [[Utility] trustless relays servicer token validation #803 (comment)](https://github.com/pokt-network/pocket/pull/803#discussion_r1228679364)
> 
>     * [[Utility] trustless relays servicer token validation #803 (comment)](https://github.com/pokt-network/pocket/pull/803#discussion_r1228705803)
> 
> 
> Will take a look at the changes shortly

Thank you for the review comments. Checked and addressed all 3.

adshmh committed:
Merge main

Olshansk commented on utility/servicer/module.go:
```diff
@@ -116,92 +124,78 @@ func (s *servicer) HandleRelay(relay *coreTypes.Relay) (*coreTypes.RelayResponse
+	case *coreTypes.Relay_RestPayload:
+		return nil, fmt.Errorf("Error executing relay on application %s: REST not supported", relay.Meta.ApplicationAddress)
+	default:
+		return nil, fmt.Errorf("Error exeucting relay on application %s: Unsupported type on payload %s", relay.Meta.ApplicationAddress, payload)
```
```suggestion
		return nil, fmt.Errorf("Error executing relay on application %s: Unsupported type on payload %s", relay.Meta.ApplicationAddress, payload)
```

adshmh replied:
Thanks, fixed.

Olshansk commented on runtime/configs/config.go:
```diff
@@ -92,15 +107,60 @@ func ParseConfig(cfgFile string) *Config {
 }
 
-// setViperDefaults this is a hacky way to set the default values for Viper so env var overrides work.
+// adjustConfigDefaults performs the necessary adjustments on the default configuration depending on the provided custom configuration.
```
Maybe `hydrateDefaultConfigs` is a better choice of words for this?

Not a blocker, but just thinking out loud

adshmh replied:
Thank you for the review. I reverted changes related to config, as they had resulted in too many lines of code being modified without fully resolving the config issue.
I opened an issue instead to track the incomplete default configuration: #858

Olshansk commented on runtime/configs/config.go:
```diff
@@ -36,7 +36,10 @@ type Config struct {
 
```
Lots of config changes here towards the end of the review. It's hard for me to evaluate if there are any issues (configs are always tricky), so I'm going to assume the ordering of things is correct and tested.

adshmh replied:
Thank you, great point! I reverted the changes, and added an `INCOMPLETE(#858)` item for default configuration instead.

Olshansk commented on runtime/configs/proto/servicer_config.proto:
```diff
@@ -5,11 +5,34 @@ package configs;
 
 option go_package = "github.com/pokt-network/pocket/runtime/configs";
 
+// TODO: Reevalute whether each utility actor should contain address/pubKey configs or if it should be shared
```
```suggestion
// TODO: Reevaluate whether each utility actor should contain address/pubKey configs or if it should be shared
```

adshmh replied:
Thank you, fixed.

Olshansk reviewed (changes requested):
@adshmh Just a few last-minute NITS. 

This PR has been open for a while and more changes are being introduced so I'm beginning to get some PR fatigue. After the last few nits are tended to, let's merge it in and follow up in a separate PR if we find issues.

adshmh committed:
Revert "Fix default configuration override"

This reverts commit f3bfe1387700a38296d89201c5e7cce3f78a63eb.

adshmh committed:
Use an INCOMPLETE item for default servicer config

adshmh committed:
Fix failing unit test

adshmh commented on persistence/local/module.go:
```diff
@@ -0,0 +1,80 @@
+package local
```
Thank you for the clarification. Will stick to opening new issues/PRs on future PRs to avoid introducing any new changes (already on that trajectory on this PR with multiple issues opened to track out-of-scope changes)

adshmh commented on shared/modules/persistence_module.go:
```diff
@@ -33,6 +36,11 @@ type PersistenceModule interface {
 	HandleDebugMessage(*messaging.DebugMessage) error
+
+	// GetLocalContext returns a local persistence context that can be used to store/retrieve node-specific, i.e. off-chain, data
+	// The module can maintain a single local context for both read and write operations: subsequent calls to GetLocalContext return
```
Thank you, fixed.

adshmh commented on persistence/local/module.go:
```diff
@@ -0,0 +1,80 @@
+	databasePath string
+}
+
+func WithLocalContextConfig(databasePath string) modules.ModuleOption {
```
Thank you, fixed.

adshmh committed:
Address review comments

adshmh committed:
Merge main

adshmh commented on utility/servicer/module_test.go:
```diff
@@ -191,15 +220,18 @@ func testRelay(editors ...relayEditor) *coreTypes.Relay {
-		Payload: &coreTypes.RelayPayload{
-			Method: "POST",
-			Data:   `{"id": 1, "jsonrpc": "2.0", method: "eth_blockNumber"}`,
+		RelayPayload: &coreTypes.Relay_JsonRpcPayload{
```
Great point, thank you! Fixed.

adshmh committed:
Fix failing unit tests

adshmh commented:
> @adshmh Just a few last-minute NITS.
> 
> This PR has been open for a while and more changes are being introduced so I'm beginning to get some PR fatigue. After the last few nits are tended to, let's merge it in and follow up in a separate PR if we find issues.

@Olshansk thank you for the reminder and multiple rounds of review. 
In the most recent commits I have tried to address the remaining review comments through minimum changes, even reverting a config change that had been introduced late in the PR and had not received thorough review.

Olshansk commented on utility/servicer/module_test.go:
```diff
@@ -213,6 +217,19 @@ func testRelayHeight(height int64) relayEditor {
 	}
 }
 
+func testEthGoerliRelay() relayEditor {
```
NICE!

Olshansk commented on utility/servicer/module_test.go:
```diff
@@ -248,6 +264,7 @@ func testServicerConfig() configs.ServicerConfig {
 		Address:   testServicer1.Address,
 		Services: map[string]*configs.ServiceConfig{
```
Thank you for taking the time to update the tests based on all my requests. Goes a long way!

Olshansk commented on utility/servicer/module.go:
```diff
@@ -465,6 +465,15 @@ func (s *servicer) executeHTTPRelay(meta *coreTypes.RelayMeta, payload *coreType
 
+// INCOMPLETE(#860): RESTful service relays: basic checks and execution through HTTP calls
+func (s *servicer) executeRESTRelay(meta *coreTypes.RelayMeta, _ *coreTypes.RESTPayload) (*coreTypes.RelayResponse, error) {
+	_, ok := s.config.Services[meta.RelayChain.Id]
+	if !ok {
```
```suggestion
	if _, ok := s.config.Services[meta.RelayChain.Id]; {
```

Olshansk replied:
@adshmh Noticed you missed this one. It's just an OPTIONAL NIT but let me know what you think about this patter

adshmh replied:
Good catch, thank you. Fixed. I think this improves readability. It is also how other places in this PR perform this check:

![image](https://github.com/pokt-network/pocket/assets/23505281/c6ec990a-2ba6-4a7c-b746-2807b59aabc8)

Olshansk commented on utility/servicer/module.go:
```diff
@@ -192,9 +192,9 @@ func (s *servicer) executeRelay(relay *coreTypes.Relay) (*coreTypes.RelayRespons
 		return s.executeHTTPRelay(relay.Meta, payload.JsonRpcPayload)
 	case *coreTypes.Relay_RestPayload:
-		return nil, fmt.Errorf("Error executing relay on application %s: REST not supported", relay.Meta.ApplicationAddress)
+		return s.executeRESTRelay(relay.Meta, payload.RestPayload)
```
My last NIT on this PR is just the naming and comments of these functions.

In the switch statement, we have:
`case *coreTypes.Relay_JsonRpcPayload:` -> `s.executeHTTPRelay(`
`case *coreTypes.Relay_RestPayload` -> `s.executeRESTRelay`

The first header is:

```go
// executeHTTPRequest performs the HTTP request that sends the relay to the chain's/service's URL.
func (s *servicer) executeHTTPRelay(meta *coreTypes.RelayMeta, payload *coreTypes.JSONRPCPayload) (*coreTypes.RelayResponse, error) {
```

The second header is:

```go
// INCOMPLETE(#860): RESTful service relays: basic checks and execution through HTTP calls
func (s *servicer) executeRESTRelay(meta *coreTypes.RelayMeta, _ *coreTypes.RESTPayload) (*coreTypes.RelayResponse, error) {
```

My suggestion is this:

```go
// executeJsonRPCRelay performs the relay for JSON-RPC payloads, sending them to the chain's/service's URL.
func (s *servicer) executeJsonRPCRelay(meta *coreTypes.RelayMeta, payload *coreTypes.JSONRPCPayload) (*coreTypes.RelayResponse, error) {
```

```go
// executeRESTRelay performs the relay for REST payloads, sending them to the chain's/service's URL.
// INCOMPLETE(#860): RESTful service relays: basic checks and execution through HTTP calls.
func (s *servicer) executeRESTRelay(meta *coreTypes.RelayMeta, _ *coreTypes.RESTPayload) (*coreTypes.RelayResponse, error) {
```

If the REST (a design pattern) payload needs to use `httpRelay` under the hood, that's fine.
If the JSON RPC (a specific protocol) leverages `httpRelay` under the hood (but could theoretically operate over gRPC or a weboscket if we wanted it to), that's fine.

Wdyt?

adshmh replied:
Thank you for the review. Updated the comments and the function names.

Olshansk reviewed (changes requested):
Just one more comment around naming & comments.

We are probably missing something, but we can merge it in and fix it later on main. We've already made A TON of progress on this!

adshmh commented:
@Olshansk sorry for the delay on this. I will address the remaining review comments within 1 day.

adshmh committed:
Address review comments

adshmh committed:
Merge main

adshmh committed:
Fix linter warning

Olshansk commented on utility/servicer/module.go:
```diff
@@ -424,26 +424,43 @@ func (s *servicer) executeHTTPRelay(meta *coreTypes.RelayMeta, payload *coreType
-	if serviceConfig.BasicAuth != nil && serviceConfig.BasicAuth.UserName != "" {
-		req.SetBasicAuth(serviceConfig.BasicAuth.UserName, serviceConfig.BasicAuth.Password)
+
+	auth := serviceConfig.BasicAuth
+	if auth != nil && auth.UserName != "" {
```
Optional NIT: single line?

adshmh replied:
Thanks, single line does read better. Fixed.

Olshansk reviewed (approved):
Just a couple of last-minute stylistic NITs but approving since it looks good to me!

adshmh committed:
Address review comments

adshmh commented:
> Just a couple of last-minute stylistic NITs but approving since it looks good to me!

Thank you for the detailed review. The 2 remaining review comments are now addressed. I will merge once CI checks pass.
